compute_tools/src/compute.rs

```diff
@@ -754,38 +755,44 @@ LIMIT 100",
             .await?;
             self.ext_remote_paths
                 .set(ext_remote_paths)
-                .expect("ext_remote_paths.set error");
+                .expect("this is the only time we set ext_remote_paths");
+            self.library_index
+                .set(library_index)
+                .expect("this is the only time we set library_index");
         }
         Ok(())
     }
 
-    pub async fn download_extension(&self, ext_name: &str) -> Result<()> {
+    pub async fn download_extension(&self, ext_name: &str, is_library: bool) -> Result<()> {
         match &self.ext_remote_storage {
             None => anyhow::bail!("No remote extension storage"),
             Some(remote_storage) => {
-                // TODO: eliminate useless LOAD Library calls to this function (if possible)
-                // not clear that we can distinguish between useful and useless
-                // library calls better than the below code
-                let ext_name = ext_name.replace(".so", "");
+                let mut real_ext_name = ext_name.to_string();
+                if is_library {
+                    real_ext_name = real_ext_name.replace(".so", "");
+                    real_ext_name =
+                        self.library_index.get().expect("oncelock err")[&real_ext_name].clone();
+                }
+
                 {
                     let mut already_downloaded_extensions =
                         self.already_downloaded_extensions.lock().expect("bad lock");
-                    if already_downloaded_extensions.contains(&ext_name) {
+                    if already_downloaded_extensions.contains(&real_ext_name) {
                         info!(
                             "extension {:?} already exists, skipping download",
                             &ext_name
                         );
                         return Ok(());
                     } else {
-                        already_downloaded_extensions.insert(ext_name.clone());
+                        already_downloaded_extensions.insert(real_ext_name.clone());
                     }
                 }
                 extension_server::download_extension(
```

Comment:
This looks a bit weird; we've already downloaded the extension (L780R787) but then we download it (L783R790)?

Reply:
We are preemptively adding `ext_name` to `already_downloaded_extensions`
maybe a more appropriate name is 
`extensions that have either already been downloaded or have already started to download`

Reply:
`download_started_extensions`